### DISCUSSION TITLE
Add a built-in clipboard manager

### DISCUSSION THREAD

**Discussion by haideralsh** [zed-industries/zed#24007] [Feature Requests]
### Check for existing issues

- [X] Completed

### Describe the feature

First, I want to say that Zed is slowly turning into one of my favourite editors ❤️  

That being said, I am migrating from a JetBrains IDE, and one missing feature in Zed that I had gotten so used to is the ability to hit `CMD` + `SHIFT` + `v` to open a built-in clipboard manager with a stack of the last items copied to the clipboard. I understand this could be achieved in the future using a plugin, but I think it would be a great quality of life feature especially for users who delete often _into_ the clipboard using either `CMD` + `x` or VIM.

Keep up the great work!

### If applicable, add mockups / screenshots to help present your vision of the feature

Mockup:

![image](https://github.com/user-attachments/assets/14565de9-cb4f-491d-811b-03ce375dbe45)

**Comment by 0x2CA** [comment]
I think many third-party tools can implement this function. Such as [Maccy](https://github.com/p0deje/Maccy)

**Comment by haideralsh** [comment]
I think the idea of having it built-in is that it only keeps buffers copied from Zed in its stack.

**Comment by andrewzielinski** [comment]
I would also love this feature and miss it

**Comment by Deco** [comment]
The UX of the JetBrains implementation is amazing.

A few notes about this feature, for any potential implementors:

 - The modal is opened by the `Ctrl+Shift+V` key combo.
 - The list of clipboard buffer entries is scrollable.
 - The list is most-recent first.
 - The top entry is selected by default and its buffer is shown in the preview area on modal open.
 - Clicking an entry line will select it and update the preview area to show that entry's buffer.
 - Navigating by arrow keys (or page up/down or home/end) will select entries (and update the preview area).
 - Pressing the `Paste` button will immediately close the modal and insert the selected entry's buffer into the current editing context.
 - Pressing the `Cancel` button or `Escape` key will close the modal without any other effects.
 - Pressing the `Enter` key will have the same effect as pressing the `Paste` button.
 - Pressing the `1`-`9` keys will have the same effect as selecting the Nth entry line from the top and pressing the `Paste` button.
 - Performing a paste action using the modal should cause that entry to be moved to the top of the internal list. (Only witnessed by future modal openings.)
 - Performing a paste action using the modal should update the system/platform clipboard contents, as if the entry's buffer was copied fresh from within a Zed editing context.
 - Undoing a paste action performed using the modal should NOT move it back to where it was in the list prior. It should still be considered most recent internally. (Only witnessed by future modal openings, as the modal would not be open at the time of undoing.)
 - Copying the same buffer twice will move its corresponding entry to the top of the internal list, rather than create duplicate identical entries. (Only witnessed by future modal openings.)
 - Each entry line has a `x` delete icon on the right that appears on hover, to allow clearing that individual entry from history (and the modal stays open).
 - There is a `Clear All` button to delete all remembered entries (and the modal stays open, but is now empty).
 - The maximum number of entries remembered is configurable.
 - The maximum size of any individual entry is configurable. Generally in MB. Oversized entries are ignored and not saved.
 - The preview area is read-only, but text can be selected and copied. (Note: New copies while the modal is open should NOT update the current modal. New copies can only be witnessed by future modal openings.)
 - The preview area should not swallow keypresses when focused, allowing `1`-`9`, `Enter`, `Escape`, and other relevant shortcuts to continue working.
 - UNCERTAIN: Whether the preview area should have a cursor that can be navigated using arrow keys, etc.
 - Only copy/cut events within Zed should be remembered. Copies/cuts outside of Zed are beyond the scope of this feature.
 - STRETCH GOAL (my idea): Pasting a buffer that was copied/cut outside of Zed into Zed should add it to Zed's clipboard history. (I've found myself wanting this while using JetBrains IDEs. The weird workaround is to paste it and then select it within the editor and copy again -- which is annoying).
 - The modal is resizable.
 - The modal size is remembered.
 - The remembered modal size is resettable.
 - The modal should be truly modal, preventing all other interaction with Zed whilst open. (I think? This is how the JetBrains one works, and there could be some strange edgecases if it were to remain open.)
 - STRETCH GOAL: Configuration option to open it as an inline suggestion popdown instead of a modal, and show what the effect of inserting a selected entry would be in the same ghostly way that AI suggestions appear.
 - UNCERTAIN: The clipboard history could be stored on the Zed client or server (or both?). On one hand, storing on the server would be handy as you could move between client machines and maintain the clipboard history. On the other hand, this sounds complicated and a potential performance issue (unnecessarily moving clipboard data over the network could hinder slow connections, for a feature that not all users might use).
 - UNCERTAIN: Should this feature be enabled by default? Or should be disabled by default and the first time the user presses Ctrl+Shift+V, it prompts them to enable it.
 - UNCERTAIN: Anything about rich clipboard formats. JetBrains IDEs support non-plaintext history entries (they have `Paste` and `Paste as plain text` buttons), but I've never used a rich format paste in a JetBrains IDE, intentionally or accidentally.

Relevant terminologies:
 - Clipboard History -- best title for the modal, and corresponding shortcut names, config names, etc.
 - Clipboard Manager -- not a good name, as it implies it somehow provides clipboard features beyond a history.
 - Buffer, Entry, Line -- I used these interchangably above, but there is a distinction.
 - Buffer: The clipboard data (should be stored as single buffer of data, chosen from the best clipboard format available at the time of copy/cut, for simplicity. I don't think there's a need to store multiple clipboard formats long-term, especially as on some platforms, some clipboard data providers, like Excel, generate particular formats on-demand.)
 - Entry: The record within Zed that holds the buffer and metadata.
 - Line: The row in the UI corresponding to the entry.

If I have time, I'll have a crack at implementing this myself. No promises, though!

**Comment by robsdedude** [comment]
Related https://github.com/zed-industries/zed/issues/4562

**Comment by jsynacek** [comment]
I'd say that these days this is better handled on the window manager layer.
